src/backend/distributed/transaction/transaction_recovery.c

```diff
@@ -159,7 +159,7 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
 
 	bool recoveryFailed = false;
 
-	int connectionFlags = 0;
+	int connectionFlags = WAIT_FOR_CONNECTION | REQUIRE_MAINTENANCE_CONNECTION;
```

Comment:
Why are we adding WAIT_FOR_CONNECTION here now? Same for lockgraph? If there's a reason, please add a code comment explaining why.

Reply:
I could say that's the core change of the PR :slightly_smiling_face: 
Before it, the maintenance connections were not waiting for the connection slot to be open, they just opened it and increased the connection counter, effectively "stealing" connections from regular queries. So, with this PR I have introduced separate "maintenance" pool, covered by `REQUIRE_MAINTENANCE_CONNECTION` flag, and force the connection to respect it by `WAIT_FOR_CONNECTION`. 
I can reflect that int the comment, if you insist, but it is a bit overkill, imo: the flags themselves are pretty self-explanatory.